### PR TITLE
Add mobile toolkit v3 URL helper

### LINKED ISSUE #90: Remove leading slashes from relative HREFs
To align with the [new v3 HREF strategy on the mobile toolkits](https://github.com/readium/go-toolkit/pull/82#discussion_r1607812022), leading slashes should not be added in relative HREFs.

See this comment: https://github.com/readium/go-toolkit/pull/82#discussion_r1609489862

### PR DESCRIPTION
**Pull request by chocolatkey**
This ports the kotlin-toolkit's V3 update's URL helper classes, to replace the current ones, which resolves #90

Other changes made include:
- Replace the mobile toolkits' custom implementation of URI templates with a more complete third-party implementation, which helps with service link matching
- Add the `ManifestTransformer` functionality from the mobile toolkit
- Moved to using native mediatypes instead of mime strings across the codebase, and added a few necessary helper functions
- Fix OPUS mimetype
- Implement the missing Locator `Text.Substring` function
- Added URL schemes, with support for cloud protocols, in preparation for cloud streaming

### CODE REVIEW

File: pkg/manifest/locator.go [hunk 1]
```diff
@@ -160,26 +188,39 @@ func TextFromJSON(rawJson map[string]interface{}) (t Text) {
 //
 // https://github.com/readium/architecture/tree/master/locators
 type Locator struct {
-	Href      string    `json:"href"`
-	Type      string    `json:"type"`
-	Title     string    `json:"title,omitempty"`
-	Locations Locations `json:"locations,omitempty"`
-	Text      Text      `json:"text,omitempty"`
+	Href      url.URL              `json:"href"`
+	MediaType *mediatype.MediaType `json:"type"`
```
Comment (mickael-menu) [inline pkg/manifest/locator.go]: Media types are required for locators, so I would not use a pointer here to enforce it.

```suggestion
	MediaType mediatype.MediaType `json:"type"`
```

File: pkg/parser/epub/parser_navdoc.go [hunk 2]
```diff
@@ -88,21 +88,21 @@ func parseLiElement(li *xmlquery.Node, filePath string) (link *manifest.Link) {
 		title = strings.TrimSpace(muchSpaceSuchWowMatcher.ReplaceAllString(first.InnerText(), " "))
 	}
 	rawHref := first.SelectAttr("href")
-	href := "#"
+	href := url.MustURLFromString("")
```
Comment (mickael-menu) [inline pkg/parser/epub/parser_navdoc.go]: I don't think an empty string is a valid URL, does it pass the validation? We used a `#` when we needed a "null" URL.

File: pkg/util/url/url.go [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+
+	"github.com/readium/go-toolkit/pkg/internal/extensions"
+	"golang.org/x/net/idna"
+)
+
+/*
+A Uniform Resource Locator.
+
+https://url.spec.whatwg.org/
+*/
```
Comment (mickael-menu) [inline pkg/util/url/url.go]:
```suggestion
// A Uniform Resource Locator.
//
// https://url.spec.whatwg.org/
```

File: pkg/util/url/url.go [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+// Create a [URL] from a Go net/url URL.
+func URLFromGo(url gurl.URL) (URL, error) {
+	if url.IsAbs() {
+		return AbsoluteURLFromGo(url)
+	} else {
+		return RelativeURLFromGo(url)
+	}
+}
+
+// Represents an absolute Uniform Resource Locator.
```
Comment (mickael-menu) [inline pkg/util/url/url.go]:
```suggestion
// Represents a relative Uniform Resource Locator.
```

File: pkg/util/url/url.go [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+/*
+A Uniform Resource Locator.
+
+https://url.spec.whatwg.org/
+*/
+type URL interface {
+	Path() string            // Decoded path segments identifying a location.
+	Filename() string        // Decoded filename portion of the URL path.
+	Extension() string       // Extension of the filename portion of the URL path.
+	Query() gurl.Values      // Returns the query parameters present in this URL.
```
Comment (mickael-menu) [inline pkg/util/url/url.go]: I would remove this API in favor of using straight `url.Raw().Query()` where needed for now. This way it is more future-proof if we need to add our own data structure representing query parameters, or if we switch to a different implementation of URL than `net/url`.

```suggestion
```

**Review by mickael-menu — COMMENTED** [review]
Great work, that looks much cleaner! 👍

File: pkg/parser/epub/parser_navdoc.go (hunk 2 above)
Comment (chocolatkey) [reply pkg/parser/epub/parser_navdoc.go]: @mickael-menu yes, it parses with Golang's URL parser:
image
Do you want me to change it back to an anchor link?

Comment (mickael-menu) [reply pkg/parser/epub/parser_navdoc.go]: You decide what you think is best. IMHO it would be more consistent as `#` is used elsewhere in the Go toolkit.

**Review by mickael-menu — APPROVED** [review]
Thanks for making the changes 🙏

**Pull request merged**